cpp/dolfinx/mesh/topologycomputation.cpp

```diff
@@ -478,92 +460,120 @@ get_local_indexing(MPI_Comm comm, const common::IndexMap& cell_map,
            graph::AdjacencyList<std::int32_t>, common::IndexMap,
            std::vector<std::int32_t>>
 compute_entities_by_key_matching(
-    MPI_Comm comm, const graph::AdjacencyList<std::int32_t>& cells,
-    const common::IndexMap& vertex_index_map,
-    const common::IndexMap& cell_index_map, mesh::CellType cell_type, int dim)
+    MPI_Comm comm,
+    std::vector<
+        std::tuple<mesh::CellType,
+                   std::shared_ptr<const graph::AdjacencyList<std::int32_t>>,
+                   std::shared_ptr<const common::IndexMap>>>
+        cell_lists,
+    const common::IndexMap& vertex_index_map, mesh::CellType entity_type,
+    int dim)
 {
   if (dim == 0)
   {
     throw std::runtime_error(
         "Cannot create vertices for topology. Should already exist.");
   }
 
+  assert(cell_dim(entity_type) == dim);
+
   // Start timer
   common::Timer timer("Compute entities of dim = " + std::to_string(dim));
 
-  // Initialize local array of entities
-  const std::int8_t num_entities_per_cell = cell_num_entities(cell_type, dim);
-
-  // For some cells, the num_vertices varies per facet (3 or 4)
-  int max_vertices_per_entity = 0;
-  for (int i = 0; i < num_entities_per_cell; ++i)
+  std::vector<std::vector<std::int32_t>> cell_type_entities(cell_lists.size());
+  std::vector<std::int32_t> cell_type_offsets = {0};
+  for (std::size_t k = 0; k < cell_lists.size(); ++k)
   {
-    max_vertices_per_entity
-        = std::max(max_vertices_per_entity,
-                   num_cell_vertices(cell_entity_type(cell_type, dim, i)));
+    mesh::CellType cell_type = std::get<0>(cell_lists[k]);
+    auto cells = std::get<1>(cell_lists[k]);
+    const std::size_t num_cells = cells->num_nodes();
+
+    for (int i = 0; i < cell_num_entities(cell_type, dim); ++i)
+    {
+      if (cell_entity_type(cell_type, dim, i) == entity_type)
+        cell_type_entities[k].push_back(i);
+    }
+    cell_type_offsets.push_back(cell_type_offsets.back()
+                                + num_cells * cell_type_entities[k].size());
   }
 
-  // Create map from cell vertices to entity vertices
-  auto e_vertices = get_entity_vertices(cell_type, dim);
+  int num_vertices_per_entity = num_cell_vertices(entity_type);
+  std::vector<std::int32_t> entity_list(cell_type_offsets.back()
+                                        * num_vertices_per_entity);
 
-  // List of vertices for each entity in each cell
-  const std::size_t num_cells = cells.num_nodes();
-  const std::size_t entity_list_shape0 = num_cells * num_entities_per_cell;
-  const std::size_t entity_list_shape1 = max_vertices_per_entity;
-  std::vector<std::int32_t> entity_list(entity_list_shape0 * entity_list_shape1,
-                                        -1);
-  for (std::size_t c = 0; c < num_cells; ++c)
+  for (std::size_t k = 0; k < cell_lists.size(); ++k)
   {
-    // Get vertices from cell
-    auto vertices = cells.links(c);
+    auto cell_type = std::get<0>(cell_lists[k]);
```

Comment:
I quite like structured binding in these cases i.e. `auto [cell_type, cells, cell_index_map] = cell_lists[k];`